### ISSUE TITLE
Usage Questions

### ISSUE THREAD

**Issue by slayton58** [dmlc/nnvm#65]
Making my general questions public so that they might help other people in the future. Adding answers from @tqchen from email (thanks Tianqi!)

Would you recommend using the C or C++ APIs? For some things, Symbol creation and composition for instance, the C API seems to be sufficient, but If I want to do things like walk the graph it seems I need to go down to C++, so I’m considering taking the time to just learn that and use it throughout. I also noted that you used the C++ API extensively in tinyflow.

**A: While to build something with respect to backend and do optimization, we do need C++11 so far.   It could be interesting to ask if we want to a bit bring more things into C API while make it minimum, if that makes thing more fun. I am open to that option.**

Now more specific questions:

1) What does the ograd variable in FGradient represent? Does ograd[i] represent the symbol that represents the gradient of n->inputs[i]?
**A: ograds[i] is the gradient to be backproped from n's ith outputs ( which is NodeEntry{n, i, 0})**

2) How exactly would I use the C API to perform a gradient pass? It requires the “ys”, “xs” and “ys_grad” attributes, but what actually are they? Do they have to be defined on the entire graph (i.e. “xs” = all inputs, “ys” = all outputs, “ys_grad” = gradients of all outputs), or on a per-node basis?
**A: TinyFlow's gradient is done in python(which calls C API) https://github.com/tqchen/tinyflow/blob/master/python/tinyflow/_base.py#L64 which helps to understand what happens**

**They are defined on the entire graph.** 
-  **The ys is the head gradient node you want to take gradient from, for example in deep nets, ys are softmax loss function.**
- **The xs are the inputs you want to take gradient from, usually it is the weights, but it could also be intermediate layers(so you get gradient backproped up to that layer)**
- **The ys_grad is usually set to ones() for loss function, it is set to gradient back proped to the next module's input, if we want to connect this with next module.**

**Comment by slayton58** [comment]
For a concrete example of 1: Imagine I have a convolution operator (no bias) that takes 2 inputs, `{"X", "weights"}` and produces a single output, `{"Y"}`. The gradient needs to take in `{"dY", "weights"}` and produces `{"dX", "weights_grad"}`.

From your explanation, `"dX" = ograds[0]`, and `"weights_grad"` would have to be handled separately.

Am I right in thinking that would look something like this?

```
NNVM_REGISTER_OP(conv)
.set_num_inputs(2)
.set_attr<FGradient>("FGradient", [](const NodePtr& n, const std::vector<NodeEntry>& ograds) {
    return std::vector<NodeEntry>{
        // Inputs to gradient are grad of 0th output (dY), and 1st input (weights)
        MakeNode("conv_grad", n->name+"_grad", {ograds[0], n->inputs[1]});
   };
});

NNVM_REGISTER_OP(conv_grad)
.set_num_inputs(2);

```

**Comment by slayton58** [comment]
The first one there is perfect - precisely what I wanted to do.
